src/hist/plot.py

```diff
@@ -214,185 +262,408 @@ def plot2d_full(
     eb_kwargs = _filter_dict(kwargs, "eb_")
     eb_kwargs.setdefault("label", "Histogram Data")
+    # Use "fmt" over "marker" to avoid UserWarning on keyword precedence
     eb_kwargs.setdefault("fmt", "o")
+    eb_kwargs.setdefault("linestyle", "none")
 
     # fit plot keyword arguments
-    label = "Fit"
-    if fit_fmt is not None:
-        for name, value, error in zip(parnames, popt, perr):
-            label += "\n  "
-            label += fit_fmt.format(name=name, value=value, error=error)
     fp_kwargs = _filter_dict(kwargs, "fp_")
-    fp_kwargs.setdefault("label", label)
+    fp_kwargs.setdefault("label", "Counts")
+
+    # bar plot keyword arguments
+    bar_kwargs = _filter_dict(kwargs, "bar_", ignore={"bar_width"})
 
     # uncertainty band keyword arguments
     ub_kwargs = _filter_dict(kwargs, "ub_")
     ub_kwargs.setdefault("label", "Uncertainty")
-    ub_kwargs.setdefault("alpha", 0.5)
 
-    # bar plot keyword arguments
-    bar_kwargs = _filter_dict(kwargs, "bar_", ignore={"bar_width"})
+    # ratio plot keyword arguments
+    rp_kwargs = _filter_dict(kwargs, "rp_")
+    rp_kwargs.setdefault("uncertainty_type", "poisson")
+    rp_kwargs.setdefault("legend_loc", "best")
+    rp_kwargs.setdefault("num_label", None)
+    rp_kwargs.setdefault("denom_label", None)
 
     # patch plot keyword arguments
-    pp_kwargs = _filter_dict(kwargs, "pp_", ignore={"pp_num"})
-    pp_num = kwargs.pop("pp_num", 5)
+    pp_kwargs = _filter_dict(kwargs, "pp_")
 
     # Judge whether some arguments are left
     if kwargs:
         raise ValueError(f"{set(kwargs)}' not needed")
 
-    # Main: plot the pulls using Matplotlib errorbar and plot methods
-    main_ax.errorbar(self.axes.centers[0], ydata, yerr, **eb_kwargs)
-
-    (line,) = main_ax.plot(self.axes.centers[0], yfit, **fp_kwargs)
-
-    # Uncertainty band
-    ub_kwargs.setdefault("color", line.get_color())
-    main_ax.fill_between(
-        self.axes.centers[0],
-        yfit - yfiterr,
-        yfit + yfiterr,
-        **ub_kwargs,
-    )
-    main_ax.legend(loc=0)
-    main_ax.set_ylabel("Counts")
-
-    # Pull: plot the pulls using Matplotlib bar method
-    left_edge = self.axes.edges[0][0]
-    right_edge = self.axes.edges[-1][-1]
-    width = (right_edge - left_edge) / len(pulls)
-    pull_ax.bar(self.axes.centers[0], pulls, width=width, **bar_kwargs)
+    main_ax.set_ylabel(fp_kwargs["label"])
 
-    patch_height = max(np.abs(pulls)) / pp_num
-    patch_width = width * len(pulls)
-    for i in range(pp_num):
-        # gradient color patches
-        if "alpha" in pp_kwargs:
-            pp_kwargs["alpha"] *= np.power(0.618, i)
+    # Computation and Fit
+    hist_values = self.values()
+
+    main_ax_artists: MainAxisArtists  # Type now due to control flow
+    if callable(other) or isinstance(other, str):
+        if isinstance(other, str):
+            if other in {"gauss", "gaus", "normal"}:
+                other = _construct_gaussian_callable(self)
```

Comment:
@henryiii I'm trying to change `_plot_ratiolike`'s `other` arg to be type `Union[hist.BaseHist, Callable[[Union[np.ndarray, float]], np.ndarray], str]` but this is now causing `mypy` to error in several places like here with things like

```
error: Incompatible types in assignment (expression has type "Callable[[ndarray, float, float, float], ndarray]", variable has type "Union[BaseHist, Callable[[Union[ndarray, float]], ndarray], str]")
```

Why is `Callable[[ndarray, float, float, float], ndarray]` not compatible with `Callable[[Union[ndarray, float]], ndarray]`? As `Callable` wants a list of types for the first argument it would seem that this should be fine. :?

Reply:
`Callable[[Union[ndarray, float]], ndarray]` is `f(_: ndarray) -> ndarray` or `f(_: float) -> ndarray`.

While `Callable[[ndarray, float, float, float], ndarray]` is f(_: ndarray, _1: float, _2: float, _3: float) -> ndarray`.

I don't see how those are compatible at all.

Reply:
Okay, when presented like this they're obviously not. Previous `mypy` errors for Callable seem to be giving confusing results IMO, but looking again at https://docs.python.org/3/library/typing.html#callable Callable requires you to specify all types of all args if you're not using `...` so I'm guessing that this instead needs to be something ugly looking in terms of Unions.

Reply:
I'd guess Union of Callables would be needed to cover defaulted arguments? This is a rather weird case since normally you call the callable with `f(x)` (in which case, you just list one) _OR_ `f(a,b,c,d)`.

Reply:
Ahh, it's due to the return type. Maybe it should be the minimum type `Callable[[np.ndarray], np.ndarray]` instead?

Reply:
That will work :+1:  and I guess as that works in `mypy` that my attempts to reflect that the differences were confused.